service-registry/src/main/java/io/servicecomb/serviceregistry/discovery/DiscoveryTree.java

```diff
@@ -0,0 +1,138 @@
+
+  private final Object lock = new Object();
+
+  private List<DiscoveryFilter> filters = new ArrayList<>();
+
+  public void loadFromSPI(Class<? extends DiscoveryFilter> cls) {
+    filters.addAll(SPIServiceUtils.getSortedService(cls));
+  }
+
+  public void addFilter(DiscoveryFilter filter) {
+    filters.add(filter);
+  }
+
+  // group name are qualifiedName
+  // all leaf group will create a loadbalancer instance, groupName is loadBalancer key
+  public void sort() {
+    filters.sort((f1, f2) -> {
+      return Integer.compare(f1.getOrder(), f2.getOrder());
+    });
+
+    LOGGER.info("found DiscoveryFilter:");
+    for (DiscoveryFilter filter : filters) {
+      LOGGER.info("DiscoveryFilter {}.", filter.getClass().getName());
+    }
+  }
+
+  protected boolean isExpired(VersionedCache existing, VersionedCache inputCache) {
+    return existing == null || existing.isExpired(inputCache);
+  }
+
+  public DiscoveryTreeNode discovery(DiscoveryContext context, String appId, String microserviceName,
+      String versionRule) {
+    VersionedCache instanceVersionedCache = RegistryUtils
+        .getServiceRegistry()
+        .getInstanceCacheManager()
+        .getOrCreateVersionedCache(appId, microserviceName, versionRule);
+
+    return discovery(context, instanceVersionedCache);
+  }
+
+  public DiscoveryTreeNode discovery(DiscoveryContext context, VersionedCache inputCache) {
+    // must save root, otherwise, maybe use old cache to create children in new root
+    DiscoveryTreeNode tmpRoot = root;
+    if (isExpired(tmpRoot, inputCache)) {
+      synchronized (lock) {
+        if (isExpired(tmpRoot, inputCache)) {
+          root = new DiscoveryTreeNode().cacheVersion(inputCache.cacheVersion());
+          tmpRoot = root;
+        }
+      }
+    }
+
+    // tmpRoot.cacheVersion() >= inputCache.cacheVersion()
+    // 1) thread 1, use v1 inputCache, did not assign tmpRoot, and suspended
+    // 2) thread 2, use v2 inputCache, update root instance
+    // 3) thread 1 go on, in this time, tmpRoot.cacheVersion() > inputCache.cacheVersion()
+    //    is not expired
+    //    but if create old children in new version root, it's a wrong logic
+    //    and this is rarely to happen, so we only let it go with a real temporary root.
+    if (tmpRoot.cacheVersion() > inputCache.cacheVersion()) {
+      tmpRoot = new DiscoveryTreeNode().cacheVersion(inputCache.cacheVersion());
+    }
+
+    DiscoveryTreeNode parent = tmpRoot.children().computeIfAbsent(inputCache.name(), name -> {
+      return new DiscoveryTreeNode().fromCache(inputCache);
+    });
+    return doDiscovery(context, parent);
+  }
+
+  protected DiscoveryTreeNode doDiscovery(DiscoveryContext context, DiscoveryTreeNode parent) {
+    for (int idx = 0; idx < filters.size();) {
+      DiscoveryFilter filter = filters.get(idx);
+      context.setCurrentNode(parent);
+
+      DiscoveryTreeNode child = filter.discovery(context, parent);
+      if (child == null) {
```

Comment:
If child is null, will generate a NullPointerException, it's the same

Reply:
yes, but from null exception, we can  not know which filter throw it.